### PR TITLE
Fix: Patch RPATHs of non-Python extension dependencies

### LINKED ISSUE #136: Non-python extension dependencies with external dependencies are not recursively patched
As reported by @thomaslima:

Our wheel has a complicated dependency structure. But the gist of it is the following:

dbcore.so (py_ext) -> lib_tl.so (non-py_ext) -> libcurl.so (external library to be grafted)

Because lib_tl is a non-py extension of dbcore, it was not added to the `full_external_refs` dictionary, which is used by `repair.py` to patch the sonames. So the final result is that lib_tl was not patched, and the package was broken.

See also #134.

### PR DESCRIPTION
**Pull request by bdice**
This PR resolves #136, and replaces the open (stalled) PR #139 from @thomaslima (closes #139).

## Summary of the problem
From my comment in #136, slightly edited for clarity:
> We have a few Python extensions that link to `libfreud.so` and `libtbb.so`, neither of which are Python extensions. `libfreud.so` is also linked to `libtbb.so`. auditwheel correctly relocates the RPATHs of our Python extensions, but does not alter the RPATH of `libfreud.so` to point to the relocated `libtbb.so`.

## Solution
Add non-Python extensions to the dictionary of `full_external_refs` in `get_wheel_elfdata` in `wheel_abi.py`.

## Context
This PR adds the fix and ports over tests from a previous PR #134 that was reverted in #190 due to tests failing (I think the failures were caused by changes in the testing infrastructure, but I'm not 100% sure). Also related to this is PR #245 which handles a related issue that @thomaslima worked on in #140.

## Testing
The `hello` package in `tests/integration/nonpy_rpath` is a minimal-ish example of such a package with a non-Python extension library with an RPATH that needs to be repaired. The tests I added are based on @thomaslima's previous work, updated to match newer conventions in the testing code for auditwheel. I think the tests are relatively complete. **I manually verified that the tests fail without the fix.**

### CODE REVIEW

File: auditwheel/wheel_abi.py
Comment (bdice) [inline auditwheel/wheel_abi.py:126]: The core of the fix is just to move this line out of the "if" statement above.

**Comment by thomaslima** [comment]
Thank you so much for providing this. I really wasn't able to work on the original PRs. I wonder if you have used any of the code that I wrote for this.

**Comment by bdice** [comment]
> Thank you so much for providing this. I really wasn't able to work on the original PRs. I wonder if you have used any of the code that I wrote for this.

Yes, I based this on your previous PRs. You can see in the commit history that I cherry-picked your previous work onto a new branch before updating the code, so that you receive credit in the git history. 👍

**Comment by mayeut** [comment]
Thanks for working on this @bdice, @thomaslima 
I will review this next week-end at the latest.

**Review by mayeut — CHANGES_REQUESTED** [review]
The functional code change is OK however, I find the test case a bit too complicated and might conflict (or more exactly, the test would become useless) with #152 

A simpler, more contained test should work as well here, maybe something like `tests/integration/testdependencies` but going one level deeper in dependencies to cover the issue would work if I understand correctly ?

**Comment by bdice** [comment]
@mayeut I agree that your suggestion for a simpler test extension would be good. I had hoped it would be good enough as-is, but you're right that won't work any more with #152. It will take me some time to make these changes -- I will try to have a simpler test extension by the end of the week. Is there a release schedule or other timeline to be aware of? I would like to get this fix into the next release if possible.

**Comment by psavery** [comment]
@bdice Thanks for working on this. I can confirm this fixes #136 for us.

Until it's in a release, we'll patch this change into the docker image we're using with cibuildwheel.

**Comment by bdice** [comment]
Update: I just finished writing a new, simpler test case. I have to do some local testing and code cleanup before re-requesting review. I anticipate having updated tests for this PR finished in the next 2 days.

**Comment by mayeut** [comment]
Thanks for the update @bdice 
> Is there a release schedule or other timeline to be aware of? I would like to get this fix into the next release if possible.

We have no release schedule for `auditwheel`. Given this fixes an issue, I'd plan to release `auditwheel` at most 1 week after merging the PR depending on my free time to do so. The other PRs are not blocker for a release.

**Comment by bdice** [comment]
@mayeut I tried to create a new test, in `tests/integration/test_nonpy_dependencies`, but it's not catching the problem correctly (the tests pass without the fix). I had to learn the Python C API and a lot of gcc linker behaviors that I haven't dealt with before, and it looks like I didn't quite get the "failing case" correct. Maybe I can use the previous test infrastructure but just use another external library that isn't `zlib`? Would that fix the conflict with #152 and be alright with you as a maintainer? Do you have a recommended library that I should try to use? I appreciate the help.

**Comment by mayeut** [comment]
@bdice,
Anything other than `zlib` and `expat` would be fine I think. You can probably try using symbols from `libxcrypt`, it won't be whitelisted again and can be grafted properly.

> Is it expected that auditwheel will use the oldest possible platform tag for the ABIs in the wheel, even if --plat is given a newer tag?

Yes, see also #281

**Pull request closed without merging**
